use copy() to change actual storage order in memory

hengji-liu opened the pull request:
Use copy() to change actual storage order in memory.
Avoid jumping in memory when iterate the float array.
Since transpose() doesn't change the order in which data is stored, the C++ code still reads the data created in vstack() order.
PS: I also tried np.matrix(list(zip(u,v,r))). But the vstack() + transpose() approach is faster.

kuy-er committed:
use copy() to change actual storage order in memory
avoid jumping in memory when iterate the float array

PorkShoulderHolder reviewed (changes requested):
Ok i see why this could be faster, but im not sure that your copy() trick works to reorder the arrays.  Try running `python mf_tests.py` in your master branch vs your hj branch (you'll have to move it out of tests and compile the source of course). Here is the output for the hj branch: 
```
iter      tr_rmse          obj
   0       0.5692   8.1194e+03
   1       0.5784   8.3805e+03
   2       0.5785   8.3832e+03
   3       0.5786   8.3829e+03
   4       0.5786   8.3836e+03
   5       0.5787   8.3835e+03
   6       0.5787   8.3818e+03
   7       0.5787   8.3845e+03
   8       0.5787   8.3828e+03
   9       0.5787   8.3822e+03
  10       0.5787   8.3816e+03
  11       0.5787   8.3815e+03
  12       0.5786   8.3809e+03
  13       0.5786   8.3800e+03
  14       0.5785   8.3777e+03
  15       0.5784   8.3766e+03
  16       0.5783   8.3745e+03
  17       0.5782   8.3727e+03
  18       0.5781   8.3709e+03
  19       0.5780   8.3680e+03
```
and here is the output from training in master:
```
iter      tr_rmse          obj
   0       0.4446   6.4932e+03
   1       0.3392   4.8285e+03
   2       0.2995   4.2912e+03
   3       0.2850   4.1181e+03
   4       0.2765   4.0215e+03
   5       0.2710   3.9538e+03
   6       0.2674   3.9117e+03
   7       0.2646   3.8808e+03
   8       0.2624   3.8575e+03
   9       0.2603   3.8377e+03
  10       0.2587   3.8220e+03
  11       0.2573   3.8098e+03
  12       0.2560   3.7969e+03
  13       0.2547   3.7879e+03
  14       0.2537   3.7802e+03
  15       0.2527   3.7729e+03
  16       0.2516   3.7611e+03
  17       0.2510   3.7614e+03
  18       0.2498   3.7511e+03
  19       0.2493   3.7501e+03
``` 
notice how in your branch the rmse doesnt change. Of course we are fitting to random data in both cases, but the rmse should still drop like it does in master. The first thing I did when writing this was also try row-major ordering when reading the numpy array but i got these same questionable results. This is my fault for not making the tests catch this, but the rmse tells me that things are broken on your branch.

PorkShoulderHolder commented:
maybe copy()'s behavior is system dependent...?

hengji-liu commented:
I added this line
`std::cout << N.u << " " << N.v << " " << N.r << std::endl;
`
to the loop in
`mf::mf_problem prob_from_sparse(const int nnx, float *X)
`
And the output convinced me that the values are read correctly by the C++ program. So I guess it might not be the problem of the copy() trick. Otherwise the C++ output would be nonsense.

I also tested on the Netflix data using the following code and the uploaded file ( [ml-100k.zip](https://github.com/PorkShoulderHolder/python-libmf/files/1472920/ml-100k.zip)
This file is u.data from the 100k data, I just renamed it to ml-100k)
```
    def setUp(self):
        users = list()
        items = list()
        ratings = list()
        with open('ml-100k') as f:
            for line in f:
                uir = line.split()
                users.append(uir[0])
                items.append(uir[1])
                ratings.append(uir[2])
        self.test_data = np.vstack((users, items, ratings)).transpose().copy()
        self.mf_engine = mf.MF()

    def test_fit(self):
        print("testing fit method")
        self.mf_engine.mf_fit(self.test_data)

```
The result is
```
iter      tr_rmse          obj
   0       1.2033   2.0704e+05
   1       0.9672   1.4828e+05
   2       0.9484   1.4038e+05
   3       0.9418   1.3577e+05
   4       0.9367   1.3246e+05
   5       0.9324   1.2972e+05
   6       0.9287   1.2750e+05
   7       0.9240   1.2526e+05
   8       0.9200   1.2341e+05
   9       0.9159   1.2178e+05
  10       0.9113   1.1997e+05
  11       0.9068   1.1843e+05
  12       0.9023   1.1698e+05
  13       0.8978   1.1557e+05
  14       0.8933   1.1420e+05
  15       0.8891   1.1296e+05
  16       0.8849   1.1167e+05
  17       0.8809   1.1063e+05
  18       0.8769   1.0951e+05
  19       0.8735   1.0859e+05.
```

PorkShoulderHolder commented:
i figured out the issue. Nothing to do with your code. Thanks for the improvement.
